File: currency-flutter/lib/wire.dart
/// Echo currency wire format v0 — Flutter side.
/// Contract: ../../spec/wire.md   Conformance: test/vectors.json
///
/// Written independently from the same prose spec as the RN implementation.
/// If this and the RN build disagree, spec/ref.py decides.
library;

import 'dart:convert';
import 'dart:typed_data';

const int kVersion = 0;
const int kNoteLen = 42;

const int _offVer = 0;
const int _offFlags = 1;
const int _offNoteId = 2;
const int _offValue = 18;
const int _offHolder = 22;
const int _offLamport = 38;

const int kFlagBearer = 0x01;
const int _flagsReserved = 0xFE;

const int kRelayTtlStart = 3;

class Note {
  final int ver;
  final int flags;
  final String noteId; // 32 hex chars
  final int value; // minor units
  final String holder; // 32 hex chars, all-zero = bearer
  final int lamport;

  const Note({
    required this.ver,
    required this.flags,
    required this.noteId,
    required this.value,
    required this.holder,
    required this.lamport,
  });

  bool get isBearer => (flags & kFlagBearer) != 0;

  /// What relay and double-spend checks key on.
  String get dedupeKey => '$noteId:$lamport';

  @override
  bool operator ==(Object other) =>
      other is Note &&
      other.ver == ver &&
      other.flags == flags &&
      other.noteId == noteId &&
      other.value == value &&
      other.holder == holder &&
      other.lamport == lamport;

  @override
  int get hashCode => Object.hash(ver, flags, noteId, value, holder, lamport);
}

/// Codes must match the RN side and spec/wire.md verbatim.
class WireError implements Exception {
  final String code;
  const WireError(this.code);
  @override
  String toString() => 'WireError($code)';
}

// ------------------------------------------------------------------ hex

Uint8List _hexToBytes(String hex) {
  final out = Uint8List(hex.length ~/ 2);
  for (var i = 0; i < out.length; i++) {
    out[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return out;
}

String _bytesToHex(Uint8List b) =>
    b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

// ------------------------------------------------------------- base64url

/// 42 bytes -> 56 chars, unpadded. Padding is stripped to match the RN side.
String toB64url(Uint8List bytes) =>
    base64Url.encode(bytes).replaceAll('=', '');

Uint8List fromB64url(String text) {
  try {
    return Uint8List.fromList(base64Url.decode(base64.normalize(text)));
  } catch (_) {
    throw const WireError('bad_length');
  }
}

// ------------------------------------------------------------ encode/decode

Uint8List encode(Note note) {
  final out = Uint8List(kNoteLen);
  final view = ByteData.sublistView(out);
  out[_offVer] = note.ver;
  out[_offFlags] = note.flags;
  out.setRange(_offNoteId, _offNoteId + 16, _hexToBytes(note.noteId));
  view.setUint32(_offValue, note.value, Endian.big);
  out.setRange(_offHolder, _offHolder + 16, _hexToBytes(note.holder));
  view.setUint32(_offLamport, note.lamport, Endian.big);
  return out;
}

Note decode(Uint8List raw) {
  if (raw.length != kNoteLen) throw const WireError('bad_length');
  final ver = raw[_offVer];
  if (ver != kVersion) throw const WireError('unsupported_version');
  final flags = raw[_offFlags];
  if (flags & _flagsReserved != 0) throw const WireError('bad_flags');
  final view = ByteData.sublistView(raw);
  return Note(
    ver: ver,
    flags: flags,
    noteId: _bytesToHex(raw.sublist(_offNoteId, _offNoteId + 16)),
    value: view.getUint32(_offValue, Endian.big),
    holder: _bytesToHex(raw.sublist(_offHolder, _offHolder + 16)),
    lamport: view.getUint32(_offLamport, Endian.big),
  );
}

// ------------------------------------------------------------ relay envelope

String wrapRelay(String b64, [int ttl = kRelayTtlStart]) => 'R|$ttl|$b64';

/// Returns null when this is a direct transfer rather than a relay envelope.
({int ttl, String b64})? parseRelay(String text) {
  if (!text.startsWith('R|')) return null;
  final first = text.indexOf('|');
  final second = text.indexOf('|', first + 1);
  if (second < 0) return null;
  return (
    ttl: int.parse(text.substring(first + 1, second)),
    b64: text.substring(second + 1),
  );
}

File: currency-rn/src/wire.ts
/**
 * Echo currency wire format v0 — React Native side.
 * Contract: ../../spec/wire.md   Conformance: ../../spec/vectors.json
 *
 * Dependency-free on purpose. React Native has no `Buffer` and no `atob`/`btoa`
 * in all runtimes, so base64url is implemented here rather than imported.
 */

export const VERSION = 0;
export const NOTE_LEN = 42;

const OFF_VER = 0;
const OFF_FLAGS = 1;
const OFF_NOTE_ID = 2;
const OFF_VALUE = 18;
const OFF_HOLDER = 22;
const OFF_LAMPORT = 38;

export const FLAG_BEARER = 0x01;
const FLAGS_RESERVED = 0xfe;

export const RELAY_TTL_START = 3;

export type Note = {
  ver: number;
  flags: number;
  note_id: string; // 32 hex chars
  value: number; // minor units
  holder: string; // 32 hex chars, all-zero = bearer
  lamport: number;
};

export type WireErrorCode =
  | 'unsupported_version'
  | 'bad_length'
  | 'bad_flags'
  | 'duplicate'
  | 'conflict'
  | 'not_holder';

export class WireError extends Error {
  // Plain field, not a TS parameter property — keeps this file runnable under
  // `node --experimental-strip-types` as well as RN's Babel transform.
  code: WireErrorCode;

  constructor(code: WireErrorCode) {
    super(code);
    this.name = 'WireError';
    this.code = code;
  }
}

// ------------------------------------------------------------------ hex

function hexToBytes(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return out;
}

function bytesToHex(b: Uint8Array): string {
  let s = '';
  for (let i = 0; i < b.length; i++) s += b[i].toString(16).padStart(2, '0');
  return s;
}

// ------------------------------------------------------------- base64url

const B64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';

export function toB64url(bytes: Uint8Array): string {
  let out = '';
  for (let i = 0; i < bytes.length; i += 3) {
    const b0 = bytes[i];
    const b1 = bytes[i + 1];
    const b2 = bytes[i + 2];
    out += B64[b0 >> 2];
    out += B64[((b0 & 0x03) << 4) | ((b1 ?? 0) >> 4)];
    if (b1 === undefined) break;
    out += B64[((b1 & 0x0f) << 2) | ((b2 ?? 0) >> 6)];
    if (b2 === undefined) break;
    out += B64[b2 & 0x3f];
  }
  return out;
}

export function fromB64url(text: string): Uint8Array {
  const clean = text.replace(/=+$/, '');
  const n = clean.length;
  const bytes: number[] = [];
  let acc = 0;
  let bits = 0;
  for (let i = 0; i < n; i++) {
    const v = B64.indexOf(clean[i]);
    if (v < 0) throw new WireError('bad_length');
    acc = (acc << 6) | v;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      bytes.push((acc >> bits) & 0xff);
    }
  }
  return new Uint8Array(bytes);
}

// ------------------------------------------------------------ encode/decode

export function encode(note: Note): Uint8Array {
  const out = new Uint8Array(NOTE_LEN);
  out[OFF_VER] = note.ver;
  out[OFF_FLAGS] = note.flags;
  out.set(hexToBytes(note.note_id), OFF_NOTE_ID);
  new DataView(out.buffer).setUint32(OFF_VALUE, note.value, false); // big-endian
  out.set(hexToBytes(note.holder), OFF_HOLDER);
  new DataView(out.buffer).setUint32(OFF_LAMPORT, note.lamport, false);
  return out;
}

export function decode(raw: Uint8Array): Note {
  if (raw.length !== NOTE_LEN) throw new WireError('bad_length');
  const ver = raw[OFF_VER];
  if (ver !== VERSION) throw new WireError('unsupported_version');
  const flags = raw[OFF_FLAGS];
  if (flags & FLAGS_RESERVED) throw new WireError('bad_flags');
  const dv = new DataView(raw.buffer, raw.byteOffset, raw.byteLength);
  return {
    ver,
    flags,
    note_id: bytesToHex(raw.subarray(OFF_NOTE_ID, OFF_NOTE_ID + 16)),
    value: dv.getUint32(OFF_VALUE, false),
    holder: bytesToHex(raw.subarray(OFF_HOLDER, OFF_HOLDER + 16)),
    lamport: dv.getUint32(OFF_LAMPORT, false),
  };
}

// ------------------------------------------------------------------ helpers

export const isBearer = (n: Note) => (n.flags & FLAG_BEARER) !== 0;

export const dedupeKey = (n: Note) => `${n.note_id}:${n.lamport}`;

export function wrapRelay(b64: string, ttl: number = RELAY_TTL_START): string {
  return `R|${ttl}|${b64}`;
}

export function parseRelay(text: string): { ttl: number; b64: string } | null {
  if (!text.startsWith('R|')) return null;
  const first = text.indexOf('|');
  const second = text.indexOf('|', first + 1);
  if (second < 0) return null;
  return {
    ttl: parseInt(text.slice(first + 1, second), 10),
    b64: text.slice(second + 1),
  };
}
